service-base/src/main/java/org/eclipse/hono/service/BasicTenantAndAuthIdExtractor.java

```diff
@@ -0,0 +1,192 @@
+            final SpanContext spanContext) {
+        Objects.requireNonNull(certificate);
+        final X500Principal x500Principal = certificate.getIssuerX500Principal();
+        final String subjectDnAuthId = certificate.getSubjectX500Principal().getName();
+        return get(x500Principal, subjectDnAuthId, spanContext);
+    }
+
+    private X509Certificate getX509Cert(final SSLSession sslSession) {
+        try {
+            final Certificate[] path = sslSession.getPeerCertificates();
+            if (path.length > 0 && path[0] instanceof X509Certificate) {
+                return (X509Certificate) path[0];
+            }
+        } catch (final SSLPeerUnverifiedException e) {
+            LOG.debug("certificate chain cannot be read: {}", e.getMessage());
+        }
+        return null;
+    }
+
+    /**
+     * Gets a pair of TenantObject and authentication identity from the given username.
+     * <p>
+     * If Hono is configured for single tenant mode, the returned tenant is {@link Constants#DEFAULT_TENANT} and
+     * <em>authId</em> is set to the given username.
+     * <p>
+     * If Hono is configured for multi tenant mode, the given username is split in two around the first occurrence of
+     * the <code>&#64;</code> sign. <em>authId</em> is then set to the first part and the tenant id is derived from the
+     * second part.
+     *
+     * @param userName The user name in the format <em>authId@tenantId</em>.
+     * @param invalidUserNameExceptionSupplier Provides the exception with which the future is failed in case the given
+     *            user name is invalid. If {@code null}, the future is failed with a NoStackTraceThrowable in that case.
+     * @param spanContext The OpenTracing context to use for tracking the operation.
+     * @return A future indicating the outcome of the operation.
+     * @see org.eclipse.hono.service.auth.device.UsernamePasswordCredentials
+     */
+    public final Future<Pair<TenantObject, String>> getFromUserName(final String userName,
+            final Supplier<Exception> invalidUserNameExceptionSupplier, final SpanContext spanContext) {
+        if (userName == null) {
+            return fail(invalidUserNameExceptionSupplier, "user name not set");
+        }
+        final String tenantId;
+        final String authId;
+        if (config.isSingleTenant()) {
+            tenantId = Constants.DEFAULT_TENANT;
+            authId = userName;
+        } else {
+            // userName is <authId>@<tenantId>
+            final String[] userComponents = userName.split("@", 2);
+            if (userComponents.length == 2) {
+                tenantId = userComponents[1];
+                authId = userComponents[0];
+            } else {
+                tenantId = null;
+                authId = null;
+            }
+        }
+        if (tenantId == null) {
+            return fail(invalidUserNameExceptionSupplier, "unsupported user name format");
+        }
```

Comment:
I have seen this code in at least three places now within Hono:
1. here
1. UsernamePasswordCredentials
1. PreSharedKeyDeviceIdentity (CoAP adapter)

I wonder if this is really necessary ...
How about extending *DeviceCredentials* with support for holding a TenantObject instead of just a tenant ID and returning a *DeviceCredentials* object here instead of a Pair/Tuple?

Reply:
I've resolved the code duplication with `UsernamePasswordCredentials`. The duplication in `PreSharedKeyDeviceIdentity` can be resolved against `UsernamePasswordCredentials` in a separate PR.
Regarding use of _DeviceCredentials_: yes, returning an object of a `TenantDetailsDeviceCredentials` class (or something like that) would be a way to make dealing with the result value more convenient (not having to use pair.one(), pair.two()) and would make it more easily extensible. Have to think about the case where there are no credentials, though (unauthenticated request). I'm preparing an update...

Reply:
@sophokles73 I've pushed an update, also changing the naming in some places (e.g. from `MqttRequestTenantAndAuthIdExtractor` to `MqttRequestTenantDetailsProvider`, its mostly about the tenant details after all).